src/Optimization/hiopKKTLinSysSparseCondensed.cpp

```diff
@@ -562,16 +559,84 @@ bool hiopKKTLinSysCondensedSparse::solveCompressed(hiopVector& rx,
   Jac_dSp_->timesVec(-1.0, dd, 1.0, dx);
 
   dyd.copyFrom(dd);
-  dyd.componentMult(*dd_pert_);
+  dyd.componentMult(*Hd_);
   dyd.axpy(-1.0, rd);
 
   nlp_->runStats.kkt.tmSolveRhsManip.stop();
 
+  return true;
+}
+
+bool hiopKKTLinSysCondensedSparse::solveCompressed(hiopVector& rx,
+                                                   hiopVector& rd,
+                                                   hiopVector& ryc,
+                                                   hiopVector& ryd,
+                                                   hiopVector& dx,
+                                                   hiopVector& dd,
+                                                   hiopVector& dyc,
+                                                   hiopVector& dyd)
+{
+  assert(nlpSp_);
+  assert(HessSp_);
+  assert(Jac_dSp_);
+  assert(0 == ryc.get_size() && "this KKT does not support equality constraints");
+
+  bool bret;
+  
+  size_type nx = rx.get_size();
+  size_type nd = rd.get_size();
+  size_type nyd = ryd.get_size();
+
+  nlp_->runStats.kkt.tmSolveRhsManip.start();
+
+  // this is rhs used by the direct solve
+  if(rhs_ == NULL) {
+    rhs_ = LinearAlgebraFactory::create_vector(nlp_->options->GetString("mem_space"), nx);
+  }
+  assert(rhs_->get_size() == nx);
+
+  nlp_->log->write("RHS KKT_SPARSE_Condensed rx: ", rx,  hovIteration);
+  nlp_->log->write("RHS KKT_SPARSE_Condensed rx: ", rd,  hovIteration);
+  nlp_->log->write("RHS KKT_SPARSE_Condensed ryc:", ryc, hovIteration);
+  nlp_->log->write("RHS KKT_SPARSE_Condensed ryd:", ryd, hovIteration);
+#if 0
+  bret = solve_compressed_direct(rx, rd, ryc, ryd, dx, dd, dyc, dyd);
+#else
+  
+  if(nullptr == krylov_mat_opr_) {
+    krylov_mat_opr_ = new hiopKKTMatVecOpr(this);
+    krylov_prec_opr_ = new hiopKKTPrecondOpr(this);
+    krylov_rhs_xdycyd_ = new hiopVectorPar(nx+nd+nyd);
+    bicgstab_ = new hiopBiCGStabSolver(nx+nd+nyd, "DEFAULT", krylov_mat_opr_, krylov_prec_opr_);
+  }
+  
+  rx.copyToStarting(*krylov_rhs_xdycyd_, 0);
+  rd.copyToStarting(*krylov_rhs_xdycyd_, nx);
+  ryd.copyToStarting(*krylov_rhs_xdycyd_, nx+nd);
+
+  bicgstab_->set_x0(0.0);
+  
+  bret = bicgstab_->solve(*krylov_rhs_xdycyd_);
+  if(!bret) {
+    if(bicgstab_->get_sol_abs_resid()>1e-8 && bicgstab_->get_sol_rel_resid()>1e-8) {
+      nlp_->log->printf(hovWarning, "%s", bicgstab_->get_convergence_info().c_str());
+    }
+    //always return true for now (later we will handle large residuals)
+    bret = true;
+  }
+
+  dx.startingAtCopyFromStartingAt(0, *krylov_rhs_xdycyd_, 0);
+  dd.startingAtCopyFromStartingAt(0, *krylov_rhs_xdycyd_, nx);
+  dyd.startingAtCopyFromStartingAt(0, *krylov_rhs_xdycyd_, nx+nd);
+  
+#endif
+
   nlp_->log->write("SOL KKT_SPARSE_Condensed dx: ", dx,  hovMatrices);
   nlp_->log->write("SOL KKT_SPARSE_Condensed dd: ", dd,  hovMatrices);
   nlp_->log->write("SOL KKT_SPARSE_Condensed dyc:", dyc, hovMatrices);
   nlp_->log->write("SOL KKT_SPARSE_Condensed dyd:", dyd, hovMatrices);
```

Comment:
I see know we only record the RHS and SOL for this method. Should we also record the matrix, e.g., the KKT matrix which is used in solve_compressed_direct?

Reply:
it's on the TODO, waiting proper resolution of CSR and COO formats...